java/vector/src/main/codegen/templates/FixedValueVectors.java

```diff
@@ -208,14 +208,21 @@ private void allocateBytes(final long size) {
    * @throws org.apache.arrow.memory.OutOfMemoryException if it can't allocate the new buffer
    */
   public void reAlloc() {
-    final long newAllocationSize = allocationSizeInBytes * 2L;
-    if (newAllocationSize > MAX_ALLOCATION_SIZE)  {
+    long baseSize  = allocationSizeInBytes;
+    final int currentBufferCapacity = data.capacity();
+    if (baseSize < (long)currentBufferCapacity) {
```

Comment:
The logic here is not very straight forward to me:
* In what case `allocationSizeInBytes` is less than `data.capacity()`? 
* Why do we want to set the new allocation to `data.capacity() * 2` instead of `allocationSizeInBytes * 2`?

Reply:
Hmm...Thanks for the explanation.

What is the semantics of transfer and why doesn't it set allocateSizeInBytes for vector B in this case?

Reply:
We saw the problem while dealing with complex JSON schema -- detailed problem description is here https://github.com/apache/arrow/pull/1097

Reply:
transfer() function is aimed at transferring the data buffer (along with ownership) of one vector to target vector of same type.

It is not clear to me if we would want to transfer more state in the function. Secondly, there could be other cases where allocationSizeInBytes < buffer capacity. So in any case we are probably better off safeguarding reAlloc() regardless.

One case could be vector reset() which resets the value of allocationSizeInBytes as well. If we reAlloc() a vector after doing reset(), I think we will run into the same problem.

Reply:
Oh, didn't realize there is another PR, thanks for the context.

```
The computation inside realloc() to determine the amount of memory to be reallocated goes wrong since it makes the decision based on allocateSizeInBytes – although this vector was created as part of transfer() from 128KB source vector, allocateSizeInBytes is still at the initial/default value of 32KB
```

Still I am a bit clear about what `transfer()` is supposed to do - Why cannot we also set `allocateSizeInBytes` when transferring the vector?

Reply:
I see. But doesn't this change cause weird `reAlloc()` effect after `reset()` though? i.e. if we reset a 128Kb double vector to 32Kb and then `reAlloc()`, it would be 256Kb instead of 64Kb, and it would also have the old value (incorrect) from 32-128Kb?

Reply:
reset() already has a problem that I think we should fix across all vector types. reset() should actually not re-initialize allocationSizeInBytes at all because reset() is typically aimed at zeroing out the buffer, resetting mutator/accessor etc -- the underlying buffer capacity still remains the same after reset.

So for your example, 128KB buffer remains a 128Kb buffer zeroed out upon reset(). On a subsequent reAlloc(), we will go to 256KB. There won't be any incorrect or garbage bits lying around on the data buffer because the entire buffer is zeroed out

Reply:
Am I missing something?